### PR TITLE
fix(group): add all selected bots as admins instead of only the first

### LINKED ISSUE #364: Adding bot administrators: only one is added when multiple are selected
## Description

When adding bot administrators, selecting multiple bots results in only one being successfully added.

## Steps to Reproduce

1. Open the "Add bot administrator" dialog
2. Select multiple bots at once
3. Submit

## Expected Result

All selected bots should be successfully added as administrators.

## Actual Result

Only one bot is successfully added; the rest are dropped.

### PR DESCRIPTION
**Pull request by yujiawei**
## Problem

In the **Add Bot Administrator** dialog (`GroupManagement`), the subscriber list allows multi-select, but the submit handler only read `selectedItems[0].uid` and issued a single `setBotAdmin` PUT. Every other selected bot was silently dropped, so picking N bots only ever added one.

## Fix

- Extract the submit logic into a small, testable pure function `submitBotAdmins(uids, setBotAdmin)` that calls the single-uid endpoint once per selected bot via `Promise.allSettled`, collecting successes and failures independently (the backend only exposes the single-uid `PUT /:group_no/bot_admin/:uid` endpoint, so a per-uid loop is the right shape — no backend change needed).
- Snapshot the selected uids before awaiting, so a late `onSelect` callback can't mutate the set mid-submit.
- On partial failure, the toast now **lists the failed uids** instead of swallowing them; full failure keeps the dialog open and surfaces the backend error message; full success closes and refreshes as before.
- `Promise.allSettled` needs the `es2020` lib, added to the package `tsconfig.json`.

## Tests

New unit tests in `GroupManagement/__tests__/botAdmins.test.ts`:
- N=3 selected → `setBotAdmin` called exactly 3 times, arg set == selected set
- 2nd uid rejects → other 2 still submitted, failed uid reported
- all reject → every failure collected

```
cd packages/dmworkbase && pnpm exec vitest run src/Components/GroupManagement/__tests__/
```
→ all passed.

## Relationship to #365

#365 is already open against this issue with the same core approach (per-uid loop + `Promise.allSettled` + the es2020 tsconfig bump), and credit to that author for landing on the right shape. This PR builds on that direction and additionally (a) extracts the logic into a pure, unit-tested function and (b) makes partial failures list the specific failed uids rather than only a count, addressing the testability/error-reporting gaps. Happy to consolidate into whichever PR maintainers prefer.

Fixes #364

### CODE REVIEW

**Review by Jerry-Xin — APPROVED** [review]
Summary: Relevance gate passed; this PR fixes an in-scope `GroupManagement` bot-admin multi-select bug and is ready to merge.

💬 Non-blocking:
- 🔵 Suggestion: [botAdmins.ts](/Users/mlamp/.openclaw/tmp/pr-review-t8cx9bm1/repo/packages/dmworkbase/src/Components/GroupManagement/botAdmins.ts:21) currently relies on `setBotAdmin` returning a rejected promise. Wrapping each call with `Promise.resolve().then(() => setBotAdmin(uid))` would also collect synchronous throws instead of aborting the whole batch.
- 🔵 Suggestion: [index.tsx](/Users/mlamp/.openclaw/tmp/pr-review-t8cx9bm1/repo/packages/dmworkbase/src/Components/GroupManagement/index.tsx:292) builds partial-failure toast text by concatenating translated copy with punctuation/counts. A dedicated i18n key with placeholders would be more locale-friendly.

✅ Highlights:
- The PR correctly changes the single submitted UID into one API call per selected bot.
- `Promise.allSettled` preserves partial success behavior and reports failed UIDs.
- The selected UID snapshot avoids selection mutation during submit.
- Focused unit tests cover success, partial failure, and all-failure cases.

Verification: I attempted the focused Vitest command and `pnpm i18n:check`, but this checkout has no installed dependencies, so `vitest` and `typescript` were unavailable. `git diff --check HEAD~1..HEAD` passed.

**Review by yujiawei — COMMENTED** [review]
# Code Review — PR #385 (octo-web)

## Summary

Solid, well-scoped bug fix. The "Add Bot Administrator" dialog is genuinely multi-select but the old handler only submitted `selectedItems[0].uid`, silently dropping every other selection. This PR correctly fans out one `setBotAdmin` PUT per selected bot, extracts the logic into a pure, unit-tested helper, and improves partial-failure reporting. The backend only exposes the single-uid `PUT /:group_no/bot_admin/:uid` endpoint (confirmed in the datasource), so the per-uid loop is the right shape — no backend change needed.

No correctness, security, or data-loss issues found. Recommending approval with a few non-blocking suggestions.

## Verification

- ✅ **Root cause is real and the fix addresses it** — `setBotAdmin(channel, uid)` is single-uid only; the snapshot + `Promise.allSettled` per-uid loop submits all selected bots.
- ✅ **Race guard is correct** — uids are snapshotted (`selectedItems.map(...)`) before the first `await`, so a late `onSelect` callback can't mutate the set mid-submit.
- ✅ **i18n keys exist** — `groupManagement.added`, `groupManagement.selectBot`, `groupManagement.operationFailed` are all present in both `zh-CN` and `en-US`.
- ✅ **tsconfig change is necessary and safe** — base config sets `target: es2019` with no `lib`, so `Promise.allSettled` types are unavailable without it. `es2020` lib is a superset of `es2019`; `Promise.allSettled` is a runtime method (not syntax), supported by all current target browsers.
- ✅ **Tests cover the meaningful branches** — N=3 all-success, single rejection (others still submitted + failed uid reported), all-reject. Order preservation is asserted.
- ✅ **Already-admin bots excluded** — the subscriber filter (`orgData?.bot_admin !== 1`) plus `disableSelectList` prevent re-adding existing admins.

## Findings

### P2 — Diverges from an existing in-repo convention for `allSettled`
This same package deliberately **avoids** `Promise.allSettled` elsewhere and hand-rolls the equivalent with `Promise.all` + per-task `.catch()`, with an explicit comment that `target=es2019` lacks the type (see `Components/Conversation/index.tsx` ~L477 and `Components/Conversation/vm.ts` ~L608). This PR instead bumps the package `lib` to `es2020`, introducing a second, inconsistent pattern for the same problem.

Both approaches are valid. The `lib` bump is the cleaner long-term direction, but it's a package-wide change made for one call site and contradicts an established local convention. Two options, either is fine:
- Keep the `lib: es2020` bump (preferred for the future) — consider a follow-up to converge the Conversation code onto `allSettled` too, so the package has one pattern.
- Or, to keep this PR minimal and consistent with neighbors, reuse the existing `Promise.all` + `.catch()` shape and drop the tsconfig change entirely.

Not blocking — just flagging the inconsistency so it's a conscious choice.

### P2 — Partial-failure toast lists raw uids, not bot names (UX)
On partial failure the toast renders `failed.map(f => f.uid).join(", ")`. UIDs are opaque identifiers; users select bots by name. Surfacing the human-readable bot name (available on the selected `Subscriber` items) would make the error actionable. Minor.

### P2 — Dialog closes on any success, discarding the selection (UX)
When some bots succeed and some fail, the code calls `context.pop()` (closing the dialog) and then shows the partial-failure toast. The user can no longer retry just the failed bots without reopening and reselecting. Keeping the dialog open on partial failure (and only closing on full success) would be friendlier. Debatable trade-off, non-blocking.

### Nit — `as any` cast on failure reason
`failed[0].reason as any` to read `.msg`. The interface correctly types `reason: unknown`; the repo guidance discourages `any`. This matches the pre-existing `err?.msg` pattern in the same file, so it's consistent — but a small typed helper (e.g. `(r: unknown) => (r as { msg?: string })?.msg`) would avoid the `any` and could be shared with the sibling `managerAdd` handler.

## Conclusion

The fix is correct, minimal, and well-tested for the extracted logic. The findings above are maintainability/UX suggestions, none of which block merge. **Approved.**

**Review by mochashanyao — APPROVED** [review]
[Octo-Q · automated review]

Verdict: Approve — no blocking findings; notes below (data-flow traced).

---

# octo-web PR#385 Review Report
**Reviewer**: Octo-Q (automated review)
**PR:** Mininglamp-OSS/octo-web#385
**Head SHA:** `7b539f642a79e0ef44e19511bb84201204628d77`
**Title:** fix(group): add all selected bots as admins instead of only the first
**Files changed:** 4 (+111 / -9)

---

## 1. Verification Summary

| Item | Status | Evidence |
|------|--------|----------|
| Bug fix correctness | ✅ | `botAdmins.ts:22-33` — `Promise.allSettled` iterates all uids, collects success/failure independently |
| Data flow: selectedItems → uids | ✅ | `index.tsx:278` — `selectedItems.map((item) => item.uid)` snapshots before await |
| Data flow: uids → setBotAdmin API | ✅ | `index.tsx:279-281` — closure captures `channel`, calls `channelDataSource.setBotAdmin(channel, uid)` per uid → `PUT groups/{channelID}/bot_admin/{uid}` |
| Race condition protection | ✅ | `index.tsx:278` — uid array captured synchronously before async work; late `onSelect` callbacks cannot mutate |
| Error handling (full success) | ✅ | `index.tsx:289-290` — toast success, dialog closed, members refreshed |
| Error handling (full failure) | ✅ | `index.tsx:291-295` — dialog stays open, error toast shows backend message |
| Error handling (partial failure) | ✅ | `index.tsx:296-302` — toast lists failed uids with count |
| Test coverage | ✅ | `botAdmins.test.ts` — 3 tests: N=3 success, partial failure, total failure |
| tsconfig lib bump | ✅ | `tsconfig.json` — adds `es2020` for `Promise.allSettled`, plus `dom`/`dom.iterable` |
| Pure function extraction | ✅ | `botAdmins.ts` — decoupled from DataSource/channel, testable in isolation |

## 2. Findings

### No P0 or P1 issues found.

### P2 Observations

**P2-1: Dialog closes on partial success (UX choice)**
- **File:** `index.tsx:283-286`
- **Diff-scope:** new (this PR introduces the partial-failure path; old code had no concept of partial success)
- **Observation:** When some bots succeed and some fail, the dialog closes (`context.pop()`) and the list refreshes. The failed uids are shown in a toast, but the user cannot retry from the dialog. This is a reasonable UX choice (successful bots are already persisted server-side, reopening the dialog would show stale state), but worth noting for future iteration.
- **Severity:** P2 — UX improvement suggestion, not a bug.

**P2-2: No test for empty uids array**
- **File:** `botAdmins.test.ts`
- **Diff-scope:** new
- **Observation:** `submitBotAdmins([], setBotAdmin)` is not tested. The component guards against this (`selectedItems.length === 0` check at `index.tsx:270`), but the pure function itself should handle it gracefully. `Promise.allSettled([])` resolves to `[]`, so the function would return `{succeeded: [], failed: []}` — correct behavior, but untested.
- **Severity:** P2 — edge case coverage gap.

**P2-3: tsconfig lib override is narrower than TypeScript defaults**
- **File:** `packages/dmworkbase/tsconfig.json`
- **Diff-scope:** new
- **Observation:** The base config (`packages/tsconfig/base.json`) has `target: "es2019"` and no explicit `lib`. TypeScript's default for es2019 includes `["es2019", "dom", "dom.iterable", "dom.asynciterable", "webworker", ...]`. The PR sets `lib: ["es2020", "dom", "dom.iterable"]`, which drops `dom.asynciterable` and webworker types. Unlikely to cause issues in a React component library, but if any file in this package uses `WebWorker` types or async iterables on DOM types, it would produce type errors.
- **Severity:** P2 — low risk, informational.

## 3. Suggestions

1. Consider adding a test for `submitBotAdmins([], fn)` to document the empty-input contract.
2. For partial failure, consider whether keeping the dialog open with pre-selected failed uids would improve UX (future iteration, not blocking).

## 4. Additional Findings

- The PR correctly identifies that the backend only exposes a single-uid endpoint (`PUT /:group_no/bot_admin/:uid`), and the per-uid loop with `Promise.allSettled` is the right client-side shape. No backend change needed.
- The `SubscriberList.onSelect` callback fires on every click (not just at submit time), but the uid snapshot at `index.tsx:278` correctly prevents any race between late `onSelect` calls and the in-flight submit.
- `submitBotAdmins` cannot throw: `Promise.allSetled` never rejects. The `await` in `index.tsx:279` is safe without a try/catch wrapper.

## 5. Data Flow Trace

```
User selects bots in SubscriberList
  → onSelect callback fires per click, updates closure var `selectedItems`
  → User clicks "Done" → onFinish handler fires
  → selectedItems.map(item => item.uid) — synchronous snapshot at index.tsx:278
  → submitBotAdmins(uids, closure) — botAdmins.ts:22
    → Promise.allSettled(uids.map(uid => setBotAdmin(uid)))
      → Each: WKApp.dataSource.channelDataSource.setBotAdmin(channel, uid)
        → WKApp.apiClient.put(`groups/${channel.channelID}/bot_admin/${uid}`)
  → Results collected: succeeded[], failed[]
  → UI response: close dialog + refresh (if any success), toast (success/error/partial)
```

All data flows verified: uids originate from `Subscriber.uid` (assigned from `memberMap.uid` in `toSubscriber`), flow through the snapshot, through `submitBotAdmins`, to the PUT endpoint. No data loss, no empty-array-at-consumption risk, no gate short-circuits.

## 6. Three Blind-Spot Checklist (R5)

- **C1 — Dual-path parity:** N/A. This PR only touches the add-bot-admin path. The remove-bot-admin path (`removeBotAdmin`) is a separate single-uid operation not affected by this change. No add↔remove symmetry issue.
- **C2 — Control-flow ordering / nested reuse:** Clear. `submitBotAdmins` is a new function called from exactly one site. No nested reuse or ordering risk.
- **C3 — Authorization boundary ≠ capability boundary:** Clear. Authorization is enforced server-side on the PUT endpoint. This client-side change does not alter who can call the endpoint or bypass any auth check.

---

**[Octo-Q] verdict: APPROVE**

Clean, well-structured fix. The core bug (only first selected bot added) is correctly resolved. The pure function extraction is testable and decoupled. Error handling covers all three paths (full success, full failure, partial failure) with appropriate user feedback. Data flow fully traced and verified. No P0/P1 issues. Three P2 observations are informational and non-blocking.

File: packages/dmworkbase/src/Components/GroupManagement/index.tsx [hunk 1]
```diff
@@ -268,17 +269,31 @@ export class GroupManagement extends Component<
...
-              uid
-            );
-            Toast.success(t("base.groupManagement.added"));
+          // 后端无批量端点，对每个选中 bot 各发一次 PUT；先快照选中 uid，
+          // 避免提交期间 onSelect 回调改写 selectedItems 造成竞态。
+          const uids = selectedItems.map((item) => item.uid);
+          const { succeeded, failed } = await submitBotAdmins(uids, (uid) =>
+            WKApp.dataSource.channelDataSource.setBotAdmin(channel, uid)
+          );
+          if (succeeded.length > 0) {
```
Comment (lml2468) [inline packages/dmworkbase/src/Components/GroupManagement/index.tsx:278]: [nit] **部分失败后对话框被关闭，无法只重试失败的 bot**

逻辑：只要 `succeeded.length > 0` 就 `context.pop()` 关闭对话框，即使有部分 uid 失败。结果是用户必须重新打开「添加 Bot 管理员」对话框、重新勾选才能重试失败的 bot，而失败提示仅以 Toast（几秒后消失）列出 uid。从交互正确性看，部分失败是最需要让用户原地纠错的场景，却恰恰丢失了已选状态。

建议：部分失败时不要 pop 对话框（或仅把成功的从选中集合移除），保留失败 uid 处于选中态，让用户可直接再次点击「确定」重试；成功的已通过 `loadMembers` 刷新到列表，不会重复添加（`disableSelectList` 已含已是管理员者）。

File: packages/dmworkbase/src/Components/GroupManagement/index.tsx [hunk 2]
```diff
@@ -268,17 +269,31 @@ export class GroupManagement extends Component<
...
+            Toast.success(t("base.groupManagement.added"));
+          } else if (succeeded.length === 0) {
+            const firstReason = failed[0].reason as any;
+            Toast.error(
+              firstReason?.msg || t("base.groupManagement.operationFailed")
+            );
+          } else {
+            // 部分失败：明确列出失败的 uid，不静默吞掉。
+            Toast.error(
+              t("base.groupManagement.operationFailed") +
```
Comment (lml2468) [inline packages/dmworkbase/src/Components/GroupManagement/index.tsx:293]: [minor] **部分失败 Toast 直接拼接原始 uid，对用户不可读且未国际化**

部分失败分支用 `t('operationFailed') + ' (n/m): ' + failed.map(f=>f.uid).join(', ')` 拼装提示。问题：(1) 展示的是 bot 的原始 uid 而非用户可识别的名称（`selectedItems` 是 Subscriber，应有 name/displayName 可用），用户看到一串 uid 无法判断是哪几个 bot；(2) `(n/m):` 这段格式与计数是手工拼接、未走 i18n，与项目其它文案的国际化方式不一致（参见同文件其余 `t(...)` 用法）。

建议：改用 `item.name`（回退到 uid）展示，并把『(失败数/总数)』做成带占位符的 i18n 文案，如 `t('base.groupManagement.partialBotAdminFailed', {failed, total, names})`。此为提示正确性/可用性问题，不影响核心提交逻辑。

File: packages/dmworkbase/src/Components/GroupManagement/index.tsx (hunk 1 above)
Comment (lml2468) [inline packages/dmworkbase/src/Components/GroupManagement/index.tsx:278]: [minor] **index.tsx 的判定/Toast 分支逻辑（本 PR 的真正风险点）完全无测试**

本 PR 把可测的纯逻辑抽成了 `submitBotAdmins` 并配了 3 个测试，但消费侧新增的判定逻辑（行 278-297）却零覆盖，而这恰恰是更容易回归的部分：
1) 「只要有成功就 `context.pop()`+`loadMembers()`」（行 278-282）；
2) 三分支 Toast：全成功 success / 全失败取 `failed[0].reason?.msg` / 部分失败拼接 `(${failed.length}/${uids.length}): ` + 失败 uid 列表（行 283-297）。

这些分支（尤其部分失败这条新路径）没有任何断言保护。建议：对 `handleAddBotAdmin`/`onFinish` 加一个轻量测试，mock `WKApp.dataSource.channelDataSource.setBotAdmin`、`context.pop`、`this.loadMembers` 与 Toast，断言：全成功→success+pop+loadMembers 各一次；部分失败→pop+loadMembers 仍调用且 `Toast.error` 文案含 `1/3` 与失败 uid；全失败→不 pop、不 loadMembers、`Toast.error` 用 `reason.msg`。若组件难以直接实例化，可把这段判定再抽成一个纯函数（输入 `{succeeded,failed,uids}`，输出 `{shouldRefresh, toast:{type,msg}}`）并单测，和 `submitBotAdmins` 抽法保持一致。

File: packages/dmworkbase/tsconfig.json [hunk 3]
```diff
@@ -1,5 +1,8 @@
 {
   "extends": "../../packages/tsconfig/react-library.json",
+  "compilerOptions": {
```
Comment (lml2468) [inline packages/dmworkbase/tsconfig.json:3]: [minor] **tsconfig `lib` 覆盖偏离 monorepo 既定基线，且与同仓库刚做过的相反决策冲突**

问题：本 PR 只为 dmworkbase 单独加 `compilerOptions.lib: ["es2020","dom","dom.iterable"]`，目的是让 `Promise.allSettled` 拿到类型。但仓库的共享基线 `packages/tsconfig/base.json` 设的是 `target: es2019`，其余 6 个兄弟包（dmworkappbot/dmworkcontacts/dmworkdatasource/dmworklogin/dmworksummary/dmworktodo）的 tsconfig 都是裸 extends、不覆盖 lib。更关键的是：同仓库的 dmworktodo 在 `packages/dmworktodo/src/ui/AssigneeEditor/index.tsx:193-195` 面对完全相同的需求（批量并发 + 部分失败不互相阻断）时，做了一个**明确相反**的架构选择——刻意不用 `Promise.allSettled`、不动 tsconfig，而是手写 `.catch` 单独兜底 + `Promise.all`，并在注释里写明理由：『本包 tsconfig target=es2019 没有 allSettled 类型，手写更稳』。本 PR 等于把刚被一个 reviewer（CR lml2468 #1088）确认过的团队约定单方面推翻，造成同一个 monorepo 内两种并存且互相矛盾的处理范式，可维护性受损。

次要技术隐患：只覆盖 `lib` 不覆盖 `target`，`target` 仍是 es2019——这会让类型系统认为 `Promise.allSettled` 可用，但语法降级仍按 es2019 走；若运行时 polyfill 未覆盖到 es2020 的 `allSettled`，会出现『类型通过但运行时 undefined』的错配（allSettled 是 runtime 方法，不靠语法降级）。

建议改法（任选其一，以保持一致性为先）：
(A) 与 dmworktodo 对齐：`botAdmins.ts` 内部改用 `uids.map(uid => setBotAdmin(uid).then(() => ({ok:true,uid})).catch(reason => ({ok:false,uid,reason})))` + `Promise.all`，语义等价但不依赖 es2020 lib，从而**删掉** tsconfig 的 lib 覆盖，零基线偏移；或
(B) 若团队确实想全仓拥抱 es2020，应把改动提到共享的 `packages/tsconfig/base.json`（统一 target+lib）而非只在 dmworkbase 局部打补丁，并同步回收 dmworktodo 里的手写兜底注释，使全仓口径一致。
推荐 (A)，改动面最小、与最近的团队决策一致。

Comment (lml2468) [inline packages/dmworkbase/tsconfig.json:3]: [minor] **tsconfig.json 的 lib 覆盖缺少 why 注释，与本包重注释风格不一致**

新增的 `compilerOptions.lib = ["es2020", "dom", "dom.iterable"]` 没有任何注释。base.json 里 target 是 es2019，而本 PR 的核心代码用到了 `Promise.allSettled`（es2020 才有），这个 lib 覆盖正是为它而加——但单看 tsconfig 完全看不出原因，后续维护者很容易把它当成可疑/多余配置而误删，进而引发编译报错。本包（以及同目录 `allowNoMention.ts` / `botAdmins.ts`）的强约定就是『为非显然的配置/逻辑写明 why』，这里却破例。

建议加一行说明，例如：`// 为 botAdmins.ts 的 Promise.allSettled 提供 es2020 类型；base target 仍是 es2019。`

File: packages/dmworkbase/src/Components/GroupManagement/botAdmins.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+// 这样本函数与具体 DataSource / channel 解耦，纯逻辑可测。
+export async function submitBotAdmins(
+  uids: string[],
+  setBotAdmin: (uid: string) => Promise<void>
+): Promise<BotAdminSubmitResult> {
+  const results = await Promise.allSettled(
+    uids.map((uid) => setBotAdmin(uid))
+  );
+  const succeeded: string[] = [];
+  const failed: { uid: string; reason: unknown }[] = [];
```
Comment (lml2468) [inline packages/dmworkbase/src/Components/GroupManagement/botAdmins.ts:25]: [nit] **failed 字段的对象形状内联重复，未复用已声明的结构**

`BotAdminSubmitResult.failed` 的元素类型 `{ uid: string; reason: unknown }` 在第 11 行（接口）和第 25 行（函数内局部变量 `const failed: { uid: string; reason: unknown }[] = []`）各写了一遍。这是同一形状的两次手写，读者需对比确认两处一致，且将来若给失败项加字段需改两处。

建议抽一个具名类型复用，例如 `export interface BotAdminFailure { uid: string; reason: unknown }`，接口与局部变量都引用它，既减少重复也让『失败项』这个概念有名字、更易读。属可选的清理项。

**Review by lml2468 — COMMENTED** [review]
## PR Review 结论 · #385

**改动概览**：4 文件 +111/-9，关联 #364。将「添加 Bot 管理员」对话框的提交逻辑从只取 `selectedItems[0]` 修正为对每个选中 bot 各发一次 PUT，并抽出可测的纯函数 `submitBotAdmins`。

**测试验证**：
- testsPass = true；无新增失败 = true（基线 57 → 57）
- build = true
- lint = true

**问题统计**：
- 阻断问题：0 条
- 小问题 / nit：6 条（已逐条贴为行内评论）
  - [nit] 部分失败后对话框被关闭，无法只重试失败的 bot（index.tsx:278）
  - [minor] 部分失败 Toast 直接拼接原始 uid，不可读且未国际化（index.tsx:293）
  - [minor] index.tsx 的判定/Toast 分支逻辑无测试覆盖（index.tsx:278-297）
  - [minor] tsconfig `lib` 覆盖偏离 monorepo 基线，且与 dmworktodo 刚做过的相反决策冲突（tsconfig.json:3）
  - [minor] tsconfig `lib` 覆盖缺少 why 注释，与本包重注释风格不一致（tsconfig.json:3）
  - [nit] `failed` 字段对象形状内联重复，未复用具名类型（botAdmins.ts:25）
- 已丢弃 2 条经核验不成立的误报。

**结论：comment-only**（仅评论不表态）。核心修复方向正确、有测试、CI 通过；但 tsconfig 基线偏移与 dmworktodo 的相反决策属团队约定层面的取舍，建议交人工裁决后再合入。

**Review by lml2468 — APPROVED** [review]
✅ APPROVE — 7b539f64。修复多选 bot 管理员只加第一个，代码质量过关：submitBotAdmins 纯函数（Promise.allSettled 全部 uid + 三态测试）、选中 uid 快照防突变、三分支 Toast。CI 全绿、0 blocker → APPROVE。⚠️ 重要协调项：#385 与 #365 是同一 bug 的两个竞争 PR，都 OPEN、都多票通过、改动高度重叠。#385 是改进版（多抽 botAdmins.ts + 测试）。两个都合会冲突——建议二选一（合 #385 关 #365），勿都 merge。🟡 fast-follow：tsconfig lib:es2020 与 dmworktodo 相反约定冲突建议统一；部分失败 Toast 拼 uid 未 i18n；部分失败关对话框丢选中；消费侧三分支+空数组边界缺测试。

**Comment by Jerry-Xin** [comment]
### ⚠️ Cross-PR coordination note (follow-up to my APPROVE)

Adding this as a coordination flag, not a code blocker — my APPROVE on `7b539f64` stands (the code is independently correct: `Promise.allSettled` over all UIDs, selected-UID snapshot to avoid mutation during submit, 3-state tests).

**#385 and #365 fix the exact same bug ("multi-select bot admins only adds the first") and both are currently OPEN with multiple approvals:**
- **#365** (`fix(group): call setBotAdmin for each selected bot...`) — touches `GroupManagement/index.tsx` + `tsconfig.json`
- **#385** (this PR) — touches the same two files **plus** extracts the logic into `botAdmins.ts` + adds unit tests

The two changes overlap heavily on `index.tsx` + `tsconfig.json`. **Merging both will conflict or leave duplicate code.** #385 is the more complete version (pure-function extraction + test coverage).

**Recommendation: pick one — merge #385 (more complete) and close #365, or vice versa. Please don't merge both.**

cc the maintainers to deconflict before merge.

**Comment by yujiawei** [comment]
Thanks @lml2468 for the detailed review — all 6 comments are valid improvements (dialog retry UX, toast i18n/readability, consumer-side branch tests, tsconfig lib baseline + comment, failed-shape reuse). Per maintainer decision merging now since all three reviews are APPROVED and the comments are nit/minor severity; tracking them as a follow-up to be addressed in a subsequent PR. Resolving threads to unblock the merge queue.
